R/driver-sql-server.R

```diff
@@ -204,3 +206,68 @@ setMethod("odbcDataType", "Microsoft SQL Server",
     )
   }
 )
+
+#' @rdname SQLServer
+#' @description
+#' ## `odbcListObjects()`
+#'
+#' This method makes tables that are synonyms visible in the Connections pane.
+# See (#221).
+#' @usage NULL
+#' @export
+`odbcListObjects.Microsoft SQL Server` <- function(connection,
+                                                   catalog = NULL,
+                                                   schema = NULL,
+                                                   name = NULL,
+                                                   type = NULL,
+                                                   ...) {
+  objects <- NextMethod(
+    object = connection,
+    catalog = catalog,
+    schema = schema,
+    name = name,
+    type = type,
+    ...
+  )
+
+  synonyms <- dbGetQuery(connection, synonyms_query(connection, catalog, schema))
+
+  if (!is.null(name)) {
+    synonyms <- synonyms[synonyms$name == name, , drop = FALSE]
+  }
+
+  rbind(
+    objects,
+    data.frame(name = synonyms$name, type = rep("table", length(synonyms$name)))
+  )
+}
+
+synonyms_query <- function(conn, catalog_name = NULL, schema_name = NULL) {
+  res <- "SELECT
+            catalog = DB.name,
+            [schema] = Sch.name,
+            name   = Syn.name
+          FROM sys.synonyms          AS Syn
+            INNER JOIN sys.schemas AS Sch
+              ON Sch.schema_id = Syn.schema_id
+            INNER JOIN sys.databases AS DB
+              ON Sch.principal_id = DB.database_id"
+
+  has_catalog <- !is.null(catalog_name)
+  has_schema <- !is.null(schema_name)
+  if (!has_catalog & !has_schema) {
+    return(paste0(res, filter_is_table))
+  }
```

Comment:
```suggestion
```

I think you can just remove this?

Reply:
Ah, went ahead and removed this  at the time but just ran into why it was there. If `!has_catalog & !has_schema` then the filter has to start with `WHERE` rather than `AND`. I'll go ahead and add a test there so it's more obvious in the future. da86cc41.